### PR TITLE
3.1 - Implements multiple Flash messages

### PR DESCRIPTION
**Pull request by HavokInspiration**
New PR to target the 3.1 branch.
As stated in my last commit on #6135 and based on the core team feedback, I stripped out everything and just kept the append strategy. I just kept the `clear` option for the `set()` method.

More details on #4619 and #6135

### CODE REVIEW

File: tests/TestCase/Controller/Component/FlashComponentTest.php [hunk 1]
```diff
@@ -97,16 +99,53 @@ public function testSet()
...
+
+        $this->Flash->set('This is a test message');
+        $expected = [
+            [
+                'message' => 'This is a test message',
+                'key' => 'flash',
+                'element' => 'Flash/default',
+                'params' => []
+                    ]
+            ];
```
Comment (markstory) [inline tests/TestCase/Controller/Component/FlashComponentTest.php]: Indentation is off.

File: tests/TestCase/Controller/Component/FlashComponentTest.php [hunk 2]
```diff
@@ -97,16 +99,53 @@ public function testSet()
...
+
+        $this->Flash->set('This is another test message', ['clear' => true]);
+        $expected = [
+            [
+                'message' => 'This is another test message',
+                'key' => 'flash',
+                'element' => 'Flash/default',
+                'params' => []
+                    ]
+            ];
```
Comment (markstory) [inline tests/TestCase/Controller/Component/FlashComponentTest.php]: Indentation is off.

**Comment by markstory** [comment]
Looks good to me. I like how much simpler this is than the last pass :clap:

File: src/View/Helper/FlashHelper.php [hunk 3]
```diff
@@ -24,6 +24,14 @@
  */
 class FlashHelper extends Helper
 {
+    /**
+     * Default config for the helper.
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'stackElement' => false
```
Comment (ADmad) [inline src/View/Helper/FlashHelper.php]: Please add description in the docblock for this new config.

**Comment by bcrowe** [comment]
Would a `stackElement` ever be more complex than a single wrapping div with a class or id? If not, I'm not too crazy on the idea of injecting a wrapping element via a config option if `stackElement`s will tend to be simple creatures. Uniquely wrapping `$this->Flash->render()` within different layouts, or even on a view-by-view basis, for the sake of styling flexibility, would seem to make more sense.

I'm also still not sold on the `clear` config option. The only case I can imagine this being useful in, is setting some flashes but then redirecting elsewhere. But, in this case, there would be an implicit contractual relationship between the two actions. And, in the case that two or more actions would ever redirect to the same action using `clear`, it could only be properly handled by checking the referrer. Sounds blah. Not sure if I'm missing any use cases here though.

Otherwise, looking good so far. :+1:

**Comment by ADmad** [comment]
> Would a stackElement ever be more complex than a single wrapping div with a class or id? If not, I'm not too crazy on the idea of injecting a wrapping element via a config option if stackElements will tend to be simple creatures. Uniquely wrapping $this->Flash->render() within different layouts, or even on a view-by-view basis, for the sake of styling flexibility, would seem to make more sense.

I agree, could do away with the `stackElement` config. 

> I'm also still not sold on the clear config option. The only case I can imagine this being useful in, is setting some flashes but then redirecting elsewhere. But, in this case, there would be an implicit contractual relationship between the two actions. And, in the case that two or more actions would ever redirect to the same action using clear, it could only be properly handled by checking the referrer. Sounds blah. Not sure if I'm missing any use cases here though.

The `clear` option doesn't add any complexity so doesn't matter if it stays. Someone might find it useful.

**Comment by HavokInspiration** [comment]
The main reason I added the ``stackElement`` config was for convenience.
Another (and probably better, actually) option would be to have it in the ``render()`` method ``$options`` argument (defaulted to ``false``). That would be more flexible and would allow to pass more parameters to the stack element :

```php
$this->Flash->render('flash', [
	'stack' => [
		'element' => 'my_element',
		'title' => 'My title',
	]
]);
```
Where ``element`` would be removed from the array and what's left given as params to the element.

Anyway, I'll remove it, and we'll see how users reacts to this. If that becomes a feature request, we'll have time to think about this.

**Comment by HavokInspiration** [comment]
I removed the stackElement option, fixed the indentation and squash.

**Comment by ADmad** [comment]
:+1:

**Comment by HavokInspiration** [comment]
I started working on the doc update (flash helpers / components & migration guide).
I just need a 3.1 branch on docs.

**Comment by ADmad** [comment]
@HavokInspiration I have created a 3.1 branch on docs repo :)

**Comment by HavokInspiration** [comment]
@ADmad Thank !

**Comment by bcrowe** [comment]
Good work @HavokInspiration. Thanks for putting up with my comments. :smile:

File: src/View/Helper/FlashHelper.php [hunk 4]
```diff
@@ -57,6 +57,9 @@ class FlashHelper extends Helper
      * ]);
      * ```
      *
+     * If you have several messages stored in the Session, each messages will be rendered in their own
```
Comment (bcrowe) [inline src/View/Helper/FlashHelper.php]: "each message will be rendered its own"

**Comment by HavokInspiration** [comment]
DocBlock correction pushed.

@bcrowe No problem. That's what OS projects are about :)

**Comment by markstory** [comment]
Looks good to me as well. I like that people have an easy way to preserve the existing behaviour if they need it.

**Comment by dereuromark** [comment]
We could think of backporting the multi flash to 2.7 and the FlashComponent there to have those two more consistent. That component was freshly introduced in 2.7 so no one is using it just yet, and even if those changes will also be BC there.

**Comment by markstory** [comment]
Sounds good @dereuromark

**Pull request merged**